src/Cassandra/ProtocolEvents/CustomTimer.cs

```diff
@@ -0,0 +1,151 @@
+﻿//
+//       Copyright (C) 2019 DataStax Inc.
+//
+//    Licensed under the Apache License, Version 2.0 (the "License");
+//    you may not use this file except in compliance with the License.
+//    You may obtain a copy of the License at
+//
+//       http://www.apache.org/licenses/LICENSE-2.0
+//
+//    Unless required by applicable law or agreed to in writing, software
+//    distributed under the License is distributed on an "AS IS" BASIS,
+//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//    See the License for the specific language governing permissions and
+//    limitations under the License.
+//
+
+using System;
+using System.Diagnostics;
+using System.Threading;
+using System.Threading.Tasks;
+using Cassandra.Tasks;
+
+namespace Cassandra.ProtocolEvents
+{
+    /// <summary>
+    /// Timer that implements a sliding window style algorithm of delay which enables the timer to keep
+    /// postponing the action trigger while the <see cref="SlideDelay"/> keeps getting invoked (up to a maximum specified delay).
+    /// It also schedules all internal code execution including the provided action with the provided task factory.
+    /// </summary>
+    internal class CustomTimer : IDisposable
+    {
+        private readonly ITimerFactory _timerFactory;
+        private readonly TimeSpan _delayIncrement;
+        private readonly TimeSpan _maxDelay;
+        private readonly Action _act;
+        private volatile ITimer _timer;
+        private long _counter;
+        private long _lastActTimestamp;
+        private volatile bool _isRunning;
+        private readonly TaskFactory _timerTaskFactory;
+
+        public CustomTimer(TaskFactory taskFactory, ITimerFactory timerFactory, TimeSpan delayIncrement, TimeSpan maxDelay, Action act)
+        {
+            if (delayIncrement > maxDelay)
+            {
+                throw new ArgumentException("delayIncrement can not be greater than maxDelay");
+            }
+
+            _timerTaskFactory = taskFactory;
+            _timerFactory = timerFactory;
+            _delayIncrement = delayIncrement;
+            _maxDelay = maxDelay;
+            _act = act;
+            _isRunning = false;
+        }
+
+        public void Dispose()
+        {
+            _timer?.Dispose();
+        }
+        
+        public void SlideDelay(bool processNow)
+        {
+            // delegate can't be async or shared exclusive scheduler is pointless
+            _timerTaskFactory.StartNew(() =>
+            {
+                var timeSpan = _delayIncrement;
+                var currentTimestamp = Stopwatch.GetTimestamp();
+
+                if (processNow)
+                {
+                    timeSpan = TimeSpan.Zero;
+                }
+                else
+                {
+                    long diffTimestamp;
+                    if (!_isRunning)
+                    {
+                        Interlocked.Exchange(ref _lastActTimestamp, currentTimestamp);
```

Comment:
Hmm I used `Interlocked` because the compiler doesn't let me add the `volatile` keyword to `long` fields. The timer implementation is already assuming that it is running on an `exclusive` scheduler (I should make that clear though)

Reply:
Basically the first dataflow block and the timer are running on the same exclusive scheduler.

Reply:
Also, the initial implementation didn't have `ConcurrentQueue` and `ConcurrentDictionary` but I switched to these when debugging an issue because I thought that it could be the reason why the issue was occuring but it turned out to be something else entirely. Do you think that we can use normal `List` and `Dictionary` within the exclusive scheduler? I think TPL puts the necessary memory barriers in place between `Tasks` so it shouldn't be an issue I think.

Reply:
> Hmm I used Interlocked because the compiler doesn't let me add the volatile

You can use `Volatile.Read()` and `Volatile.Write()` instead, CAS operations are more expensive than read/writes: http://ithare.com/infographics-operation-costs-in-cpu-clock-cycles/

Reply:
Yep after I wrote this I remembered about `Volatile.Read` and `Volatile.Write` my bad, changing it now.

Reply:
I think we can use non-concurrent collections.

Reply:
done